### DISCUSSION TITLE
Share secrets with a code executor using Docker?

### DISCUSSION THREAD

**Discussion by dynobo** [microsoft/autogen#1663] [Q&A]
Hi, I have a simple code executor agent using a docker image to run Python code:

```python
coder_agent = ConversableAgent(
    llm_config:False,
    code_execution_config={"use_docker":"my-image"})
    # ...
)
```

This code executor agent should have access to an API, which requires credentials to connect.

**How can I share such credentials with the code executor agent safely?**

I can't set the secret during docker build, because it is dynamic. (It probably also wouldn't be safe.)

I _could_ create a custom `ConversableAgent` and override the `run_code()` to ingest secrets as strings into the code string. But that doesn't feel safe, as I have little control over where the code will be printed or saved. What do you think?

In a non-AutoGen environment I would do `docker run -e "MY_SECRET=password" ...`  and use `os.environ.get("MY_SECRET")` in the code, but it seems like there is no way to pass environment variables to the docker started by AutoGen.

Are there any other options?

**Comment by ekzhu** [comment]
I think it is a good idea not to share any secrete with the executor because once a secret is in the executor's environment, it is accessible to LLM, which might be prompt-injected to leak your secret. So, if you really do share the secret with the executor, then there isn't difference between you sharing it directly via string or using environment variable. 

Instead, set up proxy servers alongside your application environment, and routing all authenticated traffic through the proxies.

Also, we currently are building new executors that can be more extendible. One example is a stateful IPython kernel executor so every code block is executed in a kernel -- no need for the LLM to rewrite the whole script each time. You can try it out using

```python
proxy = UserProxyAgent("proxy", code_executor_config={"executor": "ipython-embedded"})
```

You will need to update the Assistant Agent's system message to instruct it how to use the kernel executor. You can find an example prompt in `autogen.coding.embedded_ipython_executor.py`. 

--- Update ---

Alternative to code execution, you can also use the tool call feature, which is typically more secure because you restrict what kind of code gets run by providing LLMs with existing functions.

**Reply by dynobo** [reply]
> Alternative to code execution, you can also use the tool call feature, which is typically more secure because you restrict what kind of code gets run by providing LLMs with existing functions.

That sounds great! Is there more information on how to use this tool call feature? I have trouble finding it in the docs...

PS: Totally agree on the proxy-part. This would be most secure. However, it feels overcomplicated for scenarios, where you don't expect a malicious user (e.g. because you are the only user) and basically only try to avoid accidental leakages, e.g. via the temporary `.py` files that are generated for the code to be executed.

**Comment by ma-armenta** [comment]
Hi @dynobo 

I think this would be the best way of doing it. Check out this [notebook](https://github.com/microsoft/autogen/blob/main/notebook/agentchat_transform_messages.ipynb), there under the section `Handling Sensitive Data`. An implementation of reading environmental variables, if that is were you store your secrets, and then redacting them with just string substitution would be a quick way of going around this. Regex as in the notebook example could be also a way.

**Comment by Git-Noob123** [comment]
As @ma-armenta suggested, we can replace env secrets within `apply_transform`. On the other hand If we store our secrets somewhere else(i.e. a secured db), what we can do is we retrieve all the secrets and cache them somewhere hidden, then we do the same thing in apply_transform by simply compare and replace
